patches/server/0925-Paper-Plugins.patch

```diff
@@ -0,0 +1,5591 @@
++class SpigotPluginProviderFactory implements PluginProviderFactory<SpigotPluginProvider> {
++
++    @Override
++    public SpigotPluginProvider build(JarFile file, JarEntry config, Path source) throws Exception {
++        PluginDescriptionFile descriptionFile;
++        try (InputStream inputStream = file.getInputStream(config)) {
++            descriptionFile = new PluginDescriptionFile(inputStream);
++        } catch (IOException | YAMLException ex) {
++            throw new InvalidDescriptionException(ex);
++        }
++
++        // Copied from SimplePluginManager#loadPlugins
++        // Spigot doesn't validate the name when the config is created, and instead when the plugin is loaded.
++        // Paper plugin configuration will do these checks in config serializer instead of when this is created.
++        String name = descriptionFile.getRawName();
++        if (PluginConfigConstraints.RESERVED_KEYS.contains(name.toLowerCase(Locale.ROOT))) {
++            throw new InvalidDescriptionException("Restricted name, cannot use %s as a plugin name.".formatted(name));
++        } else if (name.indexOf(' ') != -1) {
++            throw new InvalidDescriptionException("Restricted name, cannot use 0x20 (space character) in a plugin name.");
++        }
++
++        return new SpigotPluginProvider(source, file, descriptionFile);
++    }
++}
++
+diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
+index 0a843e0afbcb1af8e2641515eb244b791b819b8c..5d563a8c9520a348b61b7122cead76aabf8d429f 100644
+--- a/src/main/java/net/minecraft/server/Main.java
++++ b/src/main/java/net/minecraft/server/Main.java
+@@ -64,6 +64,7 @@ import net.minecraft.SharedConstants;
+ public class Main {
+ 
+     private static final Logger LOGGER = LogUtils.getLogger();
++    public static Path pluginDirectory;
+ 
+     public Main() {}
+ 
+@@ -105,6 +106,18 @@ public class Main {
+             }
+ 
+             Bootstrap.bootStrap();
++            // Paper start
++            io.papermc.paper.plugin.provider.service.classloader.PaperClassloaderBytecodeModifier.init();
++            io.papermc.paper.plugin.provider.source.Providers provider = io.papermc.paper.plugin.provider.source.Providers.INSTANCE;
++            File pluginfile = (File) optionset.valueOf("plugins");
++            provider.addProviders(io.papermc.paper.plugin.provider.source.impl.DirectoryProviderSource.INSTANCE, pluginfile.toPath());
++
++            @SuppressWarnings("unchecked")
++            java.util.List<File> files = (java.util.List<File>) optionset.valuesOf("add-plugin");
++            provider.addProviders(io.papermc.paper.plugin.provider.source.impl.PluginFlagProviderSource.INSTANCE, files);
++            pluginDirectory = pluginfile.toPath();
++            io.papermc.paper.plugin.provider.service.entrypoint.ServerEntryPointHandler.INSTANCE.enter(io.papermc.paper.plugin.provider.service.entrypoint.Entrypoint.BOOTSTRAPPER);
++            // Paper end
```

Comment:
Hey, I guess the main purpose of `PluginBootstrap#boostrap` is to allow people to use "registry API" in the future with it? Why didn't you put the initialization before `Bootstrap#bootStrap`? Because inside this method occurs first NMS `Registry` static invocation which causes all registries in there to initialize and freeze.

Reply:
I'm just guessing here, but that is because those registries are not going to be supported for modification. Only the registries which are re-created and then loaded from datapacks after. The BuiltInRegistries. All the default registries are frozen in Bootstrap#bootstrap, but the built in ones are recreated later.

Reply:
> I'm just guessing here, but that is because those registries are not going to be supported for modification. Only the registries which are re-created and then loaded from datapacks after. The BuiltInRegistries. All the default registries are frozen in Bootstrap#bootstrap, but the built in ones are recreated later.
I hope Owen will answer this, but 1) there are some non-networkable registries over there which have already used in some PRs (with some sort of hacks I guess?) like: `MemoryModuleType`, `SensorType`, `Schedule`, `Activity` are in WIP Brain API, `PoiType` is in WIP Poi API. And others which we can develop APIs for in the fututre: `StatType`, loot stuff etc. 2) It'd be cool if our future "registry API" had events like "EntryRegisterEvent", so while something is about to register a plugin dev could change its properties, for example: there was the explosion resistance PR (https://github.com/PaperMC/Paper/pull/7828), we could make this "property changing" more safe and clean by allowing it only if a content hasn't registered yet.

Reply:
Sorry for the wait, yes this will probably need to be moved then. My goal is to pretty much allow most registries to be overridden especially for stuff like future brain api or what not.